elastic/kibana
# Replace spy panels by Inspector

close #16544, close #16417, close #15812, close #14642, close #13137, close #13038, close #12591, close #12552, close #5251, close #16185, close #15901, close #15897, close #13597, close #11470, close #8430, close #5116, close #13571, close #19658, close #19926

Replaces the previous spy-panels by the new Inspector (written in React). It's a contextual flyout (i.e. not limited to the visualization size anymore) to inspect data about specific elements (e.g. visualizations). The architecture of the Inspector was build flexible, so it's not limited to visualizations. Everything, that passes a list of "adapters" can open an inspector (e.g. we can make discover log it's requests in a request adapter and allow the user to view them in the inspector).

In contrast to the old spy panels the state of the inspector is not stored in the URL. It's a temporary view to inspect visualizations and not meant for sharing. If you actually used spy panels to e.g. show a table on a dashboard besides a visualization, we recommend just creating that table using the same aggregations as the visualization, and place it beside it on the dashboard.

We also move the trigger to open the Inspector out of the visualization.
You can currently open the inspector as follows:

* **On a Dashboard** - Use the context menu of the panel and hit Inspector to open it.
* **In Visualization Editor** - Use the "Inspector" button on top of the screen.

## TODOs

- [x] Finalize design with @cchaos 
  - [x] Request selector design
- [x] Exchange dashboard panel action icon (new icon already in place, just need to wait for next EUI release)
- [x] Close inspectors when navigating away from Dashboards properly
- [x] Remove artificial delays/errors for async loading
- [x] Cross-read all texts
- [x] Fix sorting (blocked by https://github.com/elastic/eui/issues/425)
- [x] Extract tabify code from courier request handler to inspector utils
-  ~Store table page size when reloading data~

## To test

To test this manually I would recommend the following setup:

* Check if the inspector can be opened in visualization editor for a classical chart
* Check if the inspector button is disabled in visualization editor mode for charts without inspector (e.g.  TSVB, Vega)
* Open the inspector on a dashboard panel (Icon should not exist for e.g. TSVB or Vega)
* I would recommend creating the following charts to see all features:
  * A regular classical chart -> data and requests should be available
  * A chart containing a terms aggregation with "Other bucket Option" -> request panel will now show two requests

## Screenshots

![Inspector data panel](https://user-images.githubusercontent.com/877229/41644988-fab51918-746f-11e8-83e4-ff6a16800a03.png)

![Inspector requests panel](https://user-images.githubusercontent.com/877229/41645004-020ad144-7470-11e8-8135-7cb541c8eb76.png)

## Follow-ups (not fixed in this PR)

- ~Properly show empty state in data table (issue with tabify)~ will be handled via tabify refactoring after merge
- ~Don't use `searchSource.history` to get last request~ I will create a separate PR after this, since it will require some searchSource changes, which I would prefer to have in a separate PR

# Dev-Docs

## Inspector

The Inspector replaces spy panels, leading to a couple of changes:

* There is no longer a `showSpyPanel` parameter to the visualize loader, because the vis internal button to open the spy panel no longer exists. To show an Inspector for an embedded visualization, use the `openInspector()` method on the visualize loader’s return value.
* Spy modes and the “spyModes” app extension no longer exist. See the [Inspector README](https://github.com/elastic/kibana/blob/master/src/ui/public/inspector/README.md) for information on how to write custom Inspector views and adapters.
